### PR TITLE
Progress toward working with subcommunicators and subdomains

### PR DESCRIPTION
**Pull request by ecoon**
Ongoing ATS work as suggested the need for "DomainSets", which are collections of meshes which represent subdomains of the global mesh.  For instance, this might be doing a solve on every "column" of cells of the mesh, then using this as a transpiration sink on the subsurface global domain.  Or it might be doing a nonlinear domain decomposition/operator splitting approach, where subdomains are solved (potentially in parallel) and then a global iteration is performed around these subdomain solves to couple them.

The following changes are introduced here:

1. Enumerated set support is added to Mesh_MSTK.  
2. DomainSets are added as a concept, and updated to allow a partitioning of the subdomain into domain sets by region.
3. MeshFactory receives a new expert parameter, "create subcommunicator", in which extracted meshes are created not on the parent mesh's Comm, but on a communicator created via MPI_Comm_Split that only exists on processes that have entities in the extracted region.  Previously, extracted meshes were required to be either Serial (e.g. all extracted entities live on 1 process) or on MPI_COMM_WORLD (where all processes had some entity -- "empty" meshes were not well supported).  This defaults to false, allowing things like "fracture" meshes to continue to live on the global communicator even when empty.
4. Key conventions are changed.  Previously derivatives were named, e.g. "dwater_content_dpressure".  This is bad because it makes it hard to split -- searching for "_d" isn't general.  Also, it became less clear what domain derivatives were on.  "dsurface-water_content_dsurface-pressure" is well posed, it isn't always obvious.  Finally, domain sets make this even more complicated, e.g. "dcolumn_0-water_content_dcolumn_0-pressure". To avoid confusion, three delimiters are introduced, and these three characters may _not_ be used in mesh or variable names.  These are:  "-" splits a domain from a variable name, "|" splits a derivative from the variable differentiated with respect to, and ":" splits a domain set name from the subdomain index, e.g. "watershed:upstream-pressure" is the "upstream" subdomain of the "watershed" domain set, the "pressure" variable.  Correspondingly, there may be "dwatershed:upstream-water_content|dwatershed:upstream-pressure", which is the corresponding derivative.

Developers should prefer to not define these themselves to make their code robust to any future changes, but instead to use the manipulators provided in Keys.hh.  So prefer to write:  

    Key domain_name = Keys::getDomainInSet("watershed", "upstream");   // not domain_name = "watershed:upstream"
    Key wc_key = Keys::getKey(domain_name, "water_content");
    Key wc_deriv = Keys::getDerivKey(wc_key, Keys::getKey(domain_name, "pressure"));  

5. PK factory and a few other places are improved to support constructing evaluators and PKs from domain sets.  So for instance, if one wants to construct a bunch of "Richards" PKs, these can be constructed from a "watershed:\*-richards" spec in the PK list, and the factory will alter the domain name and other variables to replace "*" with each subdomain name in the DomainSet "watershed".
6. Finally, checkpointing into a single file is not possible from two different meshes if they do not share the same set of processes in the process universe (e.g. MPI_Comm_Compare --> MPI_SIMILAR).  Several tests were added that demonstrated that this is quite easy to break.  As a result, a new Checkpointing mode was added to allow checkpointing by domain.  This "multi-file checkpoint" is created in a subdirectory, to allow easily bundling the checkpoint "directory" for users.  The need for this can automatically be determined by the code at runtime, so currently we do the default "single file" checkpoint, and only switch to "multi-file checkpoint" when it is needed.

Actual PKs, MPCs, and demo problems using this capability exist in ATS, but are not very mature yet.  This pull request sets the baseline work on communicators and keys to enable more work in this area.

### CODE REVIEW

File: src/state/State.cc [hunk 1]
```diff
@@ -1285,82 +1288,22 @@ void WriteVis(Visualization& vis,
...
-  HDF5_MPI checkpoint(comm, filename);
-  checkpoint.open_h5file();
-
-  // load the attributes
-  double time(0.);
-  checkpoint.readAttrReal(time, "time");
-  S.set_time(time);
+double ReadCheckpoint(State& S, const std::string& filename) {
+  bool old = false;
+  if (Keys::ends_with(filename, ".h5")) old = true;
```
Comment (lipnikov) [inline src/state/State.cc:1293]: Confused since checkpoint is always a .h5 file.

File: src/state/test/state_restart.cc [hunk 2]
```diff
@@ -24,162 +24,12 @@
 
 
 SUITE(RESTART) {
-
-  TEST(RESTART_DUMP_REQUIRED_INTERVAL) {
```
Comment (lipnikov) [inline src/state/test/state_restart.cc:28]: Why did you remove these tests? Please comment.

**Review by lipnikov — APPROVED** [review]

File: src/mesh/DomainSet.cc [hunk 3]
```diff
@@ -9,35 +9,114 @@
...
+  for (const auto& index : indices) {
+    meshes_.emplace_back(Keys::getDomainInSet(name, index));
+  }
+}
+
+
+DomainSet::DomainSet(const std::string& name,
+                     const Teuchos::RCP<const Mesh>& indexing_parent,
+                     const std::vector<std::string>& indices,
+                     const Teuchos::RCP<const Mesh>& referencing_parent,
```
[the remark this answers is not in the document]
Comment (ecoon) [reply src/mesh/DomainSet.cc:37]: Updated.  I'm open to clearer names here if you have ideas.

The "indexing" parent is the mesh whose entities are used to index the set of meshes (e.g. in a column-based structure the indexing parent is the surface mesh -- one column for each surface mesh cell).

The "referencing" parent is a mesh which may or may not exist to which the sub meshes can refer to.  In the column example, this is the subsurface mesh -- we can map each column mesh _into_ the subsurface mesh, and do so for vis, etc.

File: src/mesh/mesh_mstk/Mesh_MSTK.cc [hunk 4]
```diff
@@ -862,14 +862,16 @@ void Mesh_MSTK::extract_mstk_mesh(List_ptr src_entities,
...
       }
-    idx = 0;
     MEdge_ptr me = nullptr;
-    while ((me = (MEdge_ptr) MESH_Next_Edge(mesh_,&idx)))
-      if (ME_PType(me) == POVERLAP) {
-        MEnt_Get_AttVal(me,eparentatt,&ival,&rval,&pval);
-        MEnt_Set_AttVal(me,eparentgid_att,ME_GlobalID((MEdge_ptr)pval),0.0,
-                        NULL);
-      }
+    if (entity_dim != MREGION) { // edge parents not set on 3D extraction -- maybe they should be --etc
```
[the remark this answers is not in the document]
Comment (ecoon) [reply src/mesh/mesh_mstk/Mesh_MSTK.cc:866]: Yes, the edge parent attribute is not currently implemented, either before or after this pull request.  If they do get implemented, this new if statement should get removed.  Without the if statement it currently crashes, because eparentatt was never set.

File: src/mpc/MeshInfo.cc [hunk 5]
```diff
@@ -9,12 +9,14 @@
 
 namespace Amanzi {
 
-void MeshInfo::WriteMeshCentroids(const AmanziMesh::Mesh& mesh) {
+void MeshInfo::WriteMeshCentroids(std::string domain, const AmanziMesh::Mesh& mesh) {
```
[the remark this answers is not in the document]
Comment (ecoon) [reply src/mpc/MeshInfo.cc:12]: No, in this case I need a copy because I may change the value (see line 16).  Though I agree there are a lot of places where I used "std::string" when I should have used "const std::string&", but this isn't one of them.  I've tried to correct those as I find them.

File: src/output/HDF5_MPI.hh [hunk 6]
```diff
@@ -69,6 +69,8 @@ class HDF5_MPI {
   std::string get_tag() { return tag_; }
   void set_tag(const std::string& tag) { tag_ = tag; }
 
+  Comm_ptr_type Comm() const { return viz_comm_; }
```
[the remark this answers is not in the document]
Comment (ecoon) [reply src/output/HDF5_MPI.hh:72]: Yeah, the rest of that file was all capitalized accessors and non-standard names, e.g. Time() and setTime(), NumNodes() and setNumNodes(), so I chose to stay consistent.  It would be better to fix these all at once, and preferably in the tpetra branch where this has been rewritten to some extent.

File: src/state/Checkpoint.cc [hunk 7]
```diff
@@ -10,18 +10,22 @@ Checkpointing for state.
...
 
 namespace Amanzi {
 
-Checkpoint::Checkpoint(Teuchos::ParameterList& plist, const Comm_ptr_type& comm) :
-    IOEvent(plist),
-    comm_(comm) {
+Checkpoint::Checkpoint(Teuchos::ParameterList& plist,
+                       const State& S) :
+  IOEvent(plist),
+  old_(true)
```
[the remark this answers is not in the document]
Comment (ecoon) [reply src/state/Checkpoint.cc:27]: fixed

File: src/state/State.cc (hunk 1 above)
Comment (ecoon) [reply src/state/State.cc:1293]: Not anymore -- it may be a folder containing multiple h5 files (new style), or it may be a h5 file (old style).

Though "new" and "old" may be bad names -- for many applications there is no need to use the "new" style ever.  So maybe this should be "single file" vs "multiple file" styles...

File: src/state/Checkpoint.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (ecoon) [reply src/state/Checkpoint.cc:27]: Changing to "single_file_"

File: src/state/test/state_restart.cc (hunk 2 above)
Comment (ecoon) [reply src/state/test/state_restart.cc:28]: These tests don't test Checkpoint, they test IOEvent, which is the base class shared by Checkpoint, Vis, Observations, and RequiredTime objects.  These tests were moved to src/utils/test/utils_ioevent.cc, though staring at it again, it is possible that the coverage isn't identical -- maybe open ended isn't covered in the utils_ioevent.  Will double check.

Comment (ecoon) [reply src/state/test/state_restart.cc:28]: Yeah, they were copy/pasted/moved from the corresponding Vis test, and removed from that Vis test, but their are slight differences with these ones.  Will merge these ones into the utils test.

Comment (ecoon) [reply src/state/test/state_restart.cc:28]: done

File: src/state/State.cc (hunk 1 above)
Comment (ecoon) [reply src/state/State.cc:1293]: changed to single_file_ instead of old_ or old_style_

File: src/mpc/MeshInfo.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (lipnikov) [reply src/mpc/MeshInfo.cc:12]: Maybe better to make a copy inside the function. But I have no preference.

[the remark this answers is not in the document]
Comment (ecoon) [reply src/mpc/MeshInfo.cc:12]: So I actually prefer to copy in the signature.  This is a style choice, but one I think worth noting -- to me a signature of a function can tell a lot to the user of that function what happens in that function.  Think from the user's perspective -- they see only the signature but not the implementation.  Passing by value says "I'm going to make a copy of this thing" while passing by reference tells them I won't make a copy.  Let's assume that was some big vector, and not just a string.  If you're going to take a deep copy of the vector, telling the user that in the signature is useful information for the user.

Of course, by that logic, we would use a lot more of "const double& val" and "const double val" to tell the user more information.  There are good talks out there (Herb Sutter has a few of them) on what kind of information you're giving when you design an interface.

[the remark this answers is not in the document]
Comment (lipnikov) [reply src/mpc/MeshInfo.cc:12]: Passing by reference does not prevent us from making derivatives of this variable. The whole convention look fragile to me. But this is not critical for this PR, I'll mark it as resolved. Go ahead and merge since other reviewers are silent.
